lib/facter/simp_rsync_environments.rb

```diff
@@ -10,15 +11,61 @@
 # @author Trevor Vaughan - tvaughan@onyxpoint.com
 #
 Facter.add('simp_rsync_environments') do
-  _rsync_environments = File.join('', 'var', 'simp', 'rsync', 'environments')
+  environments = File.join('', 'var', 'simp', 'environments')
 
   confine do
-    File.directory?(_rsync_environments)
+    File.directory?(environments)
   end
 
   has_weight 1
 
   setcode do
-    Dir.glob(File.join(_rsync_environments, '*')).map{|x| x = File.basename(x)}
+    environment_hash = {}
+
+    orig_dir = Dir.pwd
+
+    Dir.chdir(environments)
+
+    Dir.glob('*/rsync').each do |env_dir|
+
+      # Using this instead of Ruby's Find for symlink support
+      share_dirs = Facter::Core::Execution.exec("find -L #{File.dirname(env_dir)} -name '.shares'").split("\n").sort
+
+      share_dirs.each do |share_dir|
+        Dir.glob(File.join(File.dirname(share_dir),'*')).each do |share|
+          dir_parts = share.split('/')
+
+          last = nil
+          dir_parts.each_with_index do |p,i|
+            if dir_parts[i+2]
+              if last
+                if last.is_a?(Hash)
+                  last[p] ||= {}
+                  last = last[p]
+                end
+              else
+                environment_hash[p] ||= {}
+                last = environment_hash[p]
+              end
+            else
+              if last.is_a?(Hash)
+                # We need to add an 'id' value to the Hash because Facter
+                # downcases all Hash keys
+                last[p] ||= {}
+                last[p]['id'] = p
+                last[p]['shares'] ||= []
+                last[p]['shares'] << dir_parts[i+1]
+              end
+
+              break
+            end
+          end
+        end
+      end
+    end
+
+    Dir.chdir(orig_dir)
+
+    environment_hash
```

Comment:
Also, the OS level in the data structure never gets and 'id' key, causing issues with logic in simp-simp.